File: fsharp/BronKerbosch/DebugOnlyTracker.fs
namespace BronKerbosch

open System.Diagnostics

// Tracks the coming and going of elements in debug builds only.
type DebugOnlyTracker<'T when 'T: equality> =
    {
#if DEBUG
      mutable individuals: 'T list  // set-like, but don't force more traits on 'T here
#else
      phantom: Unit
#endif
    }

    static member init() : DebugOnlyTracker<'T> =
        {
#if DEBUG
          individuals = []
#else
          phantom = ()
#endif
        }

    member this.Count: int =
#if DEBUG
        this.individuals.Length
#else
        failwith "Debug build only, please"
#endif

    member this.Contains(element: 'T) =
#if DEBUG
        List.contains element this.individuals
#else
        failwith "debug build only, please"
#endif

    static member Add(this: byref<DebugOnlyTracker<'T>>, element: 'T) : Unit =
        Debug.Assert(this.Contains element |> not)
#if DEBUG
        this.individuals <- element :: this.individuals
#endif

    static member Remove(this: byref<DebugOnlyTracker<'T>>, element: 'T) =
        Debug.Assert(this.Contains element)
#if DEBUG
        this.individuals <- this.individuals |> List.except [ element ]
#endif

File: csharp/BronKerbosch/DebugOnlyTracker.cs
#if DEBUG
using System.Collections.Generic;
using System.Diagnostics;
#else
using System;
#endif

namespace BronKerbosch
{
    // Tracks the coming and going of elements in debug builds only,
    // geared towards FortifiedCounter<T>.
    public sealed class DebugOnlyTracker<T>
    {
#if DEBUG
        private readonly HashSet<T> itsElements = [];

        public bool Contains(T element) => itsElements.Contains(element);

        // Returns the number of elements added and not yet removed.
        public int Add(T element)
        {
            bool added = itsElements.Add(element);
            Debug.Assert(added);
            return itsElements.Count;
        }

        // Returns the number of elements added and not yet removed.
        public int Remove(T element)
        {
            bool removed = itsElements.Remove(element);
            Debug.Assert(removed);
            return itsElements.Count;
        }
#else
        public bool Contains(T _) => throw new NotImplementedException("Debug build only, please");
        public int Add(T _) => throw new NotImplementedException("Debug build only, please");
        public int Remove(T _) => throw new NotImplementedException("Debug build only, please");
#endif
    }
}
